LanguageParser.hs

```diff
@@ -1,40 +1,414 @@
+                             P.space1 <* P.string' "as" <* P.space1)
+
+                greaterThanP :: Parser Comparison
+                greaterThanP = genericComparison Gt
+                               (positiveComparatorP <* P.space1 <*
+                               P.string' "than" <* P.space1)
+
+                greaterThanEqualsP :: Parser Comparison
+                greaterThanEqualsP = genericComparison Ge
+                                     (P.string "not" <* P.space1 <*
+                                     negativeComparatorP <* P.space1 <*
+                                     P.string' "than" <* P.space1)
+
+expressionP :: Parser Expression
+expressionP = P.try varP <|>
+              P.try constantP <|>
+              P.try sumP <|>
+              P.try productP <|>
+              P.try differenceP <|>
+              P.try quotientP <|>
+              P.try squareP <|>
+              P.try cubeP <|>
+              P.try squareRootP <|>
+              P.try twiceP <|>
+              P.try modP
+
+constantP :: Parser Expression
+constantP = P.try (constP (Constant 0) (P.string' "nothing" <* P.space)) <|>
+            P.try (genericConstant (2 ^) (W.positiveNouns ++ W.neutralNouns)) <|>
+            genericConstant (negate . (2 ^)) W.negativeNouns
+            where
+              genericConstant f l =
+                Constant . f . length <$>
+                (P.try (consBeginning *> rest l) <|> rest l)
+
+              rest l = P.space *> many (oneOfString' W.adjectives <* P.space1)
+                       <* P.space <* possesives <* oneOfString' l <* P.space
+
+              possesives = optional (oneOfString' W.possesives) <* P.space
+
+              consBeginning =
+                oneOfString' (W.articles ++ W.firstPersonPossessive ++
+                             W.secondPersonPossessive ++ W.thirdPersonPossessive)
+
+              empty = const "" <$> P.takeP Nothing 0
+
+binOp :: (Expression -> Expression -> Expression) -> String -> String ->
+         Parser Expression
+binOp con word prep =
+  liftA2 con
+  (P.string' "the" *> P.space1 *> P.string' word *> P.space1 *> P.string' prep
+  *> P.space *> expressionP) (P.string' "and" *> P.space1 *> expressionP)
+
+sumP :: Parser Expression
+sumP = binOp Sum "sum" "of"
+
+differenceP :: Parser Expression
+differenceP = binOp Difference "difference" "between"
+
+productP :: Parser Expression
+productP = binOp Product "product" "of"
+
+quotientP :: Parser Expression
+quotientP = binOp Quotient "quotient" "between"
+
+unOp :: (Expression -> Expression) -> Parser a -> Parser Expression
+unOp con word = con <$> (word *> P.space *> expressionP)
+
+squareP :: Parser Expression
+squareP = unOp Square (P.string' "the" <* P.space1 <* P.string' "square" <*
+                      P.space1 <* P.string' "of")
+
+cubeP :: Parser Expression
+cubeP = unOp Cube (P.string' "the" <* P.space1 <* P.string' "cube" <*
+                  P.space1 <* P.string' "of")
```

Comment:
You do these constant sentences in so many places. It feels like it's begging for another function, so you could do `myFunc "the cube of"` and it would appropriately break on the spaces, etc. to give a parser for that sentence so you don't need all of these hard-to-read blocks.
  
`myFunc :: String -> Parser String`

Reply:
i tried this for a while before we submitted but struggled hard :(

Reply:
Hm... Maybe leave it for now and make it an issue then? I'd think it would be
```haskell
sentenceParser :: String -> Parser String
sentenceParser s = case words s of
  x:xs -> foldl (\rest word -> rest <* P.space1 <* P.string' word) (P.string x) xs
  [x]  -> P.string' x
  []   -> P.string' ""
```

Reply:
tested the above very briefly, seems to work fine

Reply:
looks great